### PR TITLE
Compact groups in BufferGeometry

### PR DESCRIPTION
**Pull request by aardgoose**
Speculative attempt to reduce draw calls by sorting and merging groups. Equivalent to Geometry.sortFacesByMaterialIndex() for BufferGeometery.

Example use in TextBufferGeometry() which reduces draw calls from approx 2 * character count to 2 for any text string if different materials are used for face and sides. The geometry/text example illustrates this.

Requires rewritting attribute buffers and the groups list so the user can not make any assumptions about the buffer layout after use.

### CODE REVIEW

**Comment by WestLangley** [comment]
How about one of these names?

    BufferGeometry.optimizeGroups()
    BufferGeometry.optimizeGroupsByMaterialIndex()
    BufferGeometry.mergeGroups()

Personally, I prefer `mergeGroups()`. Groups having identical material indices are merged into a single group. The ordering of the groups can be arbitrary.

**Comment by aardgoose** [comment]
mergeGroups looks fine, too many ways to spell optimise.

Obviously doesn't handle indexes or interleaved buffers.

File: src/core/BufferGeometry.js [hunk 1]
```diff
@@ -837,6 +837,93 @@ BufferGeometry.prototype = Object.assign( Object.create( EventDispatcher.prototy
...
+			} else {
+
+				lastGroup = group;
+				this.groups.push( lastGroup );
+
+			}
+
+		}
+
+		console.log( 'compact groups: old count:', groups.length, 'new count:', this.groups.length );
```
Comment (Mugen87) [inline src/core/BufferGeometry.js]: For consistency reasons, please write it like this:
```js
console.log( 'THREE.BufferGeometry: Old count:', groups.length, 'new count:', this.groups.length );
```

**Review by Mugen87 — CHANGES_REQUESTED** [review]

**Review by Mugen87 — APPROVED** [review]

**Comment by Mugen87** [comment]
Please don't forget to update the docs :innocent: 

https://threejs.org/docs/#api/core/BufferGeometry

File: docs/api/core/BufferGeometry.html [hunk 2]
```diff
@@ -274,6 +274,12 @@ <h3>[method:BufferGeometry lookAt] ( [param:Vector3 vector] )</h3>
 		<h3>[method:null merge]( [param:BufferGeometry bufferGeometry], [param:Integer offset] )</h3>
 		<div>Merge in another BufferGeometry with an optional offset of where to start merging in.</div>
 
+		<h3>[method:BufferGeometry mergeGroups]()</h3>
+		<div>Merge groups with common materialIndex values to minimize the number of draw calls required to
+		render the geometery. This will reorder the elements of all attributes.
+		Indexed BufferGeometries or those using an [page:InterleavedBufferAttribute] are not supported.
```
Comment (WestLangley) [inline docs/api/core/BufferGeometry.html]: I think you should test for these cases, then, and issue a console warning.

Comment (aardgoose) [reply docs/api/core/BufferGeometry.html]: I had wondered about that.
Will do.

File: src/core/BufferGeometry.js [hunk 3]
```diff
@@ -837,6 +837,97 @@ BufferGeometry.prototype = Object.assign( Object.create( EventDispatcher.prototy
...
+			} else {
+
+				lastGroup = group;
+				this.groups.push( lastGroup );
+
+			}
+
+		}
+
+		console.log( 'THREE.BufferGeometry: old count:', groups.length, 'new count:', this.groups.length );
```
Comment (WestLangley) [inline src/core/BufferGeometry.js]: Maybe leave this line, but comment it out?

**Comment by aardgoose** [comment]
I had missed that TextBufferGeometry is actually indexed with a 1:1 mapping,  reordering the index OR the other attributes works equally well in optimising draw calls.

 For an indexed geometry just reorder the index values.

(is there a reason for the 1:1 index created by ExtrudeGeometry() ?)

**Comment by WestLangley** [comment]
>(is there a reason for the 1:1 index created by ExtrudeGeometry() ?)

/ping @Mugen87 Remove the index?

**Comment by WestLangley** [comment]
>Still, a 1:1 index seems useless to me.

Sorting faces is easier with an indexed geometry.

I am thinking that a better approach here would be to sort the index if it is an indexed geometry, and add an index if it is not.

**Comment by WestLangley** [comment]
>So BufferGeometry.mergeGroups() would add an index if it is not present?

Yes, rather than sorting all the attributes. Just an idea...

**Comment by Mugen87** [comment]
Umm, I think I would be surprised if there is suddenly an index in my geometry. From my point of view, it feels more predictable if `BufferGeometry.mergeGroups()` does not do this.

**Comment by WestLangley** [comment]
Does this code work if there are vertices (or indices) that are not in any group?

**Comment by aardgoose** [comment]
vertices/indices not in any group will be ignored.  I cant see a sensible thing to do with them other than return early with a warning if the group total count is less than the total number of vertices/indices.

**Comment by WestLangley** [comment]
>vertices/indices not in any group will be ignored

That is what I thought... vertices, at least. So, I don't think this PR is acceptable.

I think this can be made to work if you add an index if it is not present -- and sort the index. If you don't want to do that, then you will have to find a way to retain the lost vertices.

If you don't like either approach, then I suggest fixing `TextBufferGeometry` so it is not so inefficient.

**Comment by aardgoose** [comment]
A few options,  1 count the number of vertices in groups and if less than vertices in position etc, either return or just attempt to merge adjacent groups with matching materials if any.  

Or 2.  as you suggest add an index and sort that, to keep to the principal of least surprise. The addition of an index could be restricted by the use of a "forceIndex" parameter which would default to false.

(that leaves the potential corner case of overlapping groups, but is it reasonable for that to be supported or have a predictable behaviour).

**Pull request closed without merging**